src/main/java/com/microsoft/azure/proton/transport/proxy/impl/ProxyImpl.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+package com.microsoft.azure.proton.transport.proxy.impl;
+
+import com.microsoft.azure.proton.transport.proxy.Proxy;
+import com.microsoft.azure.proton.transport.proxy.ProxyHandler;
+
+import org.apache.qpid.proton.engine.Transport;
+import org.apache.qpid.proton.engine.TransportException;
+import org.apache.qpid.proton.engine.impl.TransportImpl;
+import org.apache.qpid.proton.engine.impl.TransportInput;
+import org.apache.qpid.proton.engine.impl.TransportLayer;
+import org.apache.qpid.proton.engine.impl.TransportOutput;
+import org.apache.qpid.proton.engine.impl.TransportWrapper;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+import static org.apache.qpid.proton.engine.impl.ByteBufferUtils.newWriteableBuffer;
+
+public class ProxyImpl implements Proxy, TransportLayer {
+    private final int proxyHandshakeBufferSize = 4 * 1024; // buffers used only for proxy-handshake
+    private final ByteBuffer inputBuffer;
+    private final ByteBuffer outputBuffer;
+
+    private boolean tailClosed = false;
+    private boolean headClosed = false;
+    private boolean isProxyConfigured;
+    private String host = "";
+    private Map<String, String> headers = null;
+    private TransportImpl underlyingTransport;
+    private ProxyState proxyState = ProxyState.PN_PROXY_NOT_STARTED;

```

Comment:
This is just a question:
What are we doing after the state machine ended up in FAILURE state? We would never go back to NOT_STARTED, is that right? What is the requirement for the user to start again?

Reply:
That's right - @zolvarga. We wouldn't go back to NOT_STARTED. 
Intention was to signal - that this transport is closed.

Reply:
The flow goes like this: 
 * once we set `TransportImpl` to `FAILED` state - proton - invokes `onTransportError()` - with the response from `Proxy` - and closes the transport. 
 * Then, developer can make a choice based on the response - whether to retry - by recreating the connection or simply throw to the layers above.